core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java

```diff
@@ -43,32 +43,118 @@
+ *
+ * When the request is not sliced, this task is the only task created, and starts an action to perform search requests.
+ *
+ * When the request is sliced, this task can either represent a parent coordinating task (using {@link BulkByScrollTask#setParent(int)}) or
+ * a child task that performs search queries (using {@link BulkByScrollTask#setChild(Integer, float)}).
+ *
+ * We don't always know if this task will be a parent or child task when it's created, because if slices is set to "auto" it may
+ * be either depending on the number of shards in the source indices. We figure that out when the request is handled and set it on this
+ * class with {@link #setParent(int)} or {@link #setChild(Integer, float)}.
  */
-public abstract class BulkByScrollTask extends CancellableTask {
+public class BulkByScrollTask extends CancellableTask {
+
+    private ParentBulkByScrollWorker parentWorker;
+    private ChildBulkByScrollWorker childWorker;
+
     public BulkByScrollTask(long id, String type, String action, String description, TaskId parentTaskId) {
         super(id, type, action, description, parentTaskId);
     }
 
+    @Override
+    public BulkByScrollTask.Status getStatus() {
+        if (isParent()) {
+            return parentWorker.getStatus();
+        }
+
+        if (isChild()) {
+            return childWorker.getStatus();
+        }
+
+        return emptyStatus();
+    }
+
+    private BulkByScrollTask.Status emptyStatus() {
+        return new Status(Collections.emptyList(), getReasonCancelled());
+    }
+
     /**
-     * The number of sub-slices that are still running. {@link WorkingBulkByScrollTask} will always have 0 and
-     * {@link ParentBulkByScrollTask} will return the number of waiting tasks. Used to decide how to perform rethrottling.
+     * Returns true if this task is a parent of other sliced tasks. False otherwise
      */
-    public abstract int runningSliceSubTasks();
+    public boolean isParent() {
+        return parentWorker != null;
+    }
 
     /**
-     * Apply the {@code newRequestsPerSecond}.
+     * Sets this task to be a parent task for {@code slices} sliced subtasks
      */
-    public abstract void rethrottle(float newRequestsPerSecond);
+    public void setParent(int slices) {
+        if (isParent()) {
+            throw new IllegalStateException("Parent worker is already set");
+        }
+        if (isChild()) {
+            throw new IllegalStateException("Worker is already set as child");
+        }
 
-    /*
-     * Overridden to force children to return compatible status.
+        parentWorker = new ParentBulkByScrollWorker(this, slices);
+    }
+
+    /**
+     * Returns the worker object that tracks the state of sliced subtasks. Throws IllegalStateException if this task is not set to be
+     * a parent task.
      */
-    public abstract BulkByScrollTask.Status getStatus();
+    public ParentBulkByScrollWorker getParentWorker() {
```

Comment:
> I'm not really sure how module loading works, but if it uses a different classloader than core I think it would disallow package-scope access in this case.

Yeah. I think it is rude to have the same package in two modules....

>Sort of related: would it make sense to move some of the stuff covered in this PR that's in core to the reindex module?

I wouldn't worry too much about it for now. I believe @rjernst is working on something that'll let us more everything over to reindex's module again which'd be lovely.